pyramid/csrf.py

```diff
@@ -0,0 +1,226 @@
+            event['get_csrf_token'] = partial(csrf.get_csrf_token, request)
+
+
+def get_csrf_token(request):
+    """ Get the currently active CSRF token for the request passed, generating
+    a new one using ``new_csrf_token(request)`` if one does not exist. This
+    calls the equivalent method in the chosen CSRF protection implementation.
+
+    .. versionadded :: 1.8a1
+    """
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.get_csrf_token(request)
+
+
+def new_csrf_token(request):
+    """ Generate a new CSRF token for the request passed and persist it in an
+    implementation defined manner. This calls the equivalent method in the
+    chosen CSRF protection implementation.
+
+    .. versionadded :: 1.8a1
+    """
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.new_csrf_token(request)
+
+
+def check_csrf_token(request,
+                     token='csrf_token',
+                     header='X-CSRF-Token',
+                     raises=True):
+    """ Check the CSRF token returned by the :meth:`pyramid.interfaces.ICSRF`
+    implementation against the value in ``request.POST.get(token)`` (if a POST
+    request) or ``request.headers.get(header)``. If a ``token`` keyword is not
+    supplied to this function, the string ``csrf_token`` will be used to look
+    up the token in ``request.POST``. If a ``header`` keyword is not supplied
+    to this function, the string ``X-CSRF-Token`` will be used to look up the
+    token in ``request.headers``.
+
+    If the value supplied by post or by header doesn't match the value supplied
+    by ``impl.get_csrf_token()`` (where ``impl`` is an implementation of
+    :meth:`pyramid.interfaces.ICSRF`), and ``raises`` is ``True``, this
+    function will raise an :exc:`pyramid.exceptions.BadCSRFToken` exception. If
+    the values differ and ``raises`` is ``False``, this function will return
+    ``False``.  If the CSRF check is successful, this function will return
+    ``True`` unconditionally.
+
+    See :ref:`auto_csrf_checking` for information about how to secure your
+    application automatically against CSRF attacks.
+
+    .. versionadded:: 1.4a2
+
+    .. versionchanged:: 1.7a1
+       A CSRF token passed in the query string of the request is no longer
+       considered valid. It must be passed in either the request body or
+       a header.
+
+    .. versionchanged:: 1.8a1
+       Moved from pyramid.session to pyramid.csrf
+    """
+    supplied_token = ""
+    # If this is a POST/PUT/etc request, then we'll check the body to see if it
+    # has a token. We explicitly use request.POST here because CSRF tokens
+    # should never appear in an URL as doing so is a security issue. We also
+    # explicitly check for request.POST here as we do not support sending form
+    # encoded data over anything but a request.POST.
+    if token is not None:
```

Comment:
I would recommend swapping the check for the token in the headers/body. `request.POST` is more expensive than checking `request.headers`.